🌊 Streams: Make tests platform agnostic

flash1293 opened the pull request:
Fixes https://github.com/elastic/streams-program/issues/29

This PR makes the streams API tests platform agnostic.

Some changes besides basic moving over were required, documented in code

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

flash1293 committed:
move tests over

flash1293 commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/index_templates/generate_index_template.ts:
```diff
@@ -27,7 +27,7 @@ export function generateIndexTemplate(id: string) {
     data_stream: {
       hidden: false,
-      failure_store: true,
+      failure_store: isServerless ? undefined : true, // TODO: Enable failure store for serverless once it is rolled out
```
ℹ️ Failure store is not available yet on serverless, but should be soon

flash1293 commented on x-pack/test/common/utils/server_route_repository/create_supertest_service_from_repository.ts:
```diff
@@ -98,7 +98,10 @@ export function getApiClientFromSupertest<TServerRouteRepository extends ServerR
-      const headers: Record<string, string> = { 'kbn-xsrf': 'foo' };
+      const headers: Record<string, string> = {
+        'kbn-xsrf': 'foo',
+        'x-elastic-internal-origin': 'kibana',
```
Had to add that in a central spot so we can also test internal APIs

dmlemeshko replied:
Both `kbn-xsrf` and `x-elastic-internal-origin` will be assigned when you pass `withInternalHeaders: true,` in 

```
const supertestAdmin = await st.getSupertestWithRoleScope('admin', {
    useCookieHeader: true,
    withInternalHeaders: true,
  });
```

flash1293 commented on x-pack/test/api_integration/deployment_agnostic/apis/observability/streams/classic.ts:
```diff
@@ -94,10 +96,12 @@ export default function ({ getService }: FtrProviderContext) {
-          policy: 'logs',
-          type: 'ilm',
-        },
+        lifecycle: isServerless
```
ℹ️ On serverless, all streams will use dlm because ilm is not available

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/267860)
* Commit: fd658686d2964290ab6c6ca0fb29a4953fa6550c

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/267860#0194700c-fe60-4004-af7d-b14829a82e88)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267860","number":267860,"commit":"fd658686d2964290ab6c6ca0fb29a4953fa6550c"}],"number":267860}
buildkite-pr-comment-->

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/agnostic-tests

flash1293 committed:
fix linting issues

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/268043)
* Commit: f9de8ed5432889025d6f9ec7af2ebff85ba54aaa

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/268043#01947417-e842-4a85-a7bd-d563561e4277)

### History
* :broken_heart: [Build #267860](https://buildkite.com/elastic/kibana-pull-request/builds/267860) failed fd658686d2964290ab6c6ca0fb29a4953fa6550c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268043","number":268043,"commit":"f9de8ed5432889025d6f9ec7af2ebff85ba54aaa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267860","number":267860,"commit":"fd658686d2964290ab6c6ca0fb29a4953fa6550c"}],"number":268043}
buildkite-pr-comment-->

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/agnostic-tests

dmlemeshko commented on x-pack/test/api_integration/deployment_agnostic/apis/observability/streams/assets/dashboard.ts:
```diff
@@ -7,9 +7,9 @@
 
-export default function ({ getService }: FtrProviderContext) {
+export default function ({ getService }: DeploymentAgnosticFtrProviderContext) {
   const supertest = getService('supertest');
```
The main idea of deployment-agnostic tests is to avoid using `supertest` service because it is authenticated with `operator` user, which is not available for customers. It is explicitly mentioned in the [docs](https://github.com/elastic/kibana/tree/main/x-pack/test/api_integration/deployment_agnostic#tests-design-requirements)

It seems like you are testing Public APIs (e.g. `/api/streams/_status`) and the correct way is to generate API keys with realm role scoped privileges. 

```
API Authentication in Kibana: Public vs. Internal APIs

Kibana provides both public and internal APIs, each requiring authentication with the correct privileges. However, the method of testing these APIs varies, depending on how they are untilized by end users.

Public APIs: When testing HTTP requests to public APIs, API key-based authentication should be used.
It reflect how end user call these APIs. Due to existing restrictions, we utilize Admin user credentials to
generate API keys for various roles. While the API key permissions are correctly scoped according to
the assigned role, the user will internally be recognized as Admin during authentication.

Internal APIs: Direct HTTP requests to internal APIs are generally not expected. However,
for testing purposes, authentication should be performed using the Cookie header.
This approach simulates client-side behavior during browser interactions, mirroring
how internal APIs are indirectly invoked.
```

dmlemeshko replied:
@flash1293 I just noticed we have it mentioned in migration guide: https://github.com/elastic/kibana/tree/main/x-pack/test/api_integration/deployment_agnostic#migrating-existing-tests

flash1293 replied:
Thanks for the ping @dmlemeshko , will adjust

dmlemeshko reviewed (commented):
I believe `apiClient` used in tests has to be updated in favor of SAML authentication / API keys.

flash1293 committed:
remove supertest usage

flash1293 commented:
As @dmlemeshko mentioned, these tests should not use `supertest`. I changed it to use `roleScopedSupertest` with the admin role, as this is the expected usage in production - only an admin user can use streams, both from the browser UI and via direct API access.

Type failures are caused by https://github.com/elastic/kibana/pull/207062 and should be fixed soon.

flash1293 committed:
remove exclusion

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/268100)
* Commit: a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/268100#01947497-3fbc-432a-a4c8-62f820c1f262)

### History
* :broken_heart: [Build #268043](https://buildkite.com/elastic/kibana-pull-request/builds/268043) failed f9de8ed5432889025d6f9ec7af2ebff85ba54aaa
* :broken_heart: [Build #267860](https://buildkite.com/elastic/kibana-pull-request/builds/267860) failed fd658686d2964290ab6c6ca0fb29a4953fa6550c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268100","number":268100,"commit":"a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268043","number":268043,"commit":"f9de8ed5432889025d6f9ec7af2ebff85ba54aaa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267860","number":267860,"commit":"fd658686d2964290ab6c6ca0fb29a4953fa6550c"}],"number":268100}
buildkite-pr-comment-->

flash1293 committed:
Merge branch 'main' into flash1293/agnostic-tests

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/268167)
* Commit: dd14269140e717496fa5f0e4da73c5ce3b27b297

### Failed CI Steps
* [FTR Configs #96](https://buildkite.com/elastic/kibana-pull-request/builds/268167#0194751b-e89a-4008-9fb3-20df98e5a91c)
* [FTR Configs #96](https://buildkite.com/elastic/kibana-pull-request/builds/268167#0194753f-8ef0-4595-abe2-add43efc33d0)

### History
* :broken_heart: [Build #268100](https://buildkite.com/elastic/kibana-pull-request/builds/268100) failed a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c
* :broken_heart: [Build #268043](https://buildkite.com/elastic/kibana-pull-request/builds/268043) failed f9de8ed5432889025d6f9ec7af2ebff85ba54aaa
* :broken_heart: [Build #267860](https://buildkite.com/elastic/kibana-pull-request/builds/267860) failed fd658686d2964290ab6c6ca0fb29a4953fa6550c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268167","number":268167,"commit":"dd14269140e717496fa5f0e4da73c5ce3b27b297"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268100","number":268100,"commit":"a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268043","number":268043,"commit":"f9de8ed5432889025d6f9ec7af2ebff85ba54aaa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267860","number":267860,"commit":"fd658686d2964290ab6c6ca0fb29a4953fa6550c"}],"number":268167}
buildkite-pr-comment-->

dmlemeshko commented on x-pack/test/common/utils/server_route_repository/create_supertest_service_from_repository.ts:
```diff
@@ -44,49 +44,15 @@ type RepositorySupertestReturnOf<
+  st: ReturnType<typeof RoleScopedSupertestProvider>
+): Promise<RepositorySupertestClient<TServerRouteRepository>> {
+  const supertestAdmin = await st.getSupertestWithRoleScope('admin', {
+    useCookieHeader: true,
```
question: will `supertestAdmin` be used to test for private APIs? If so, all good.

But if the APIs in tests are public you need to remove this line and then API key will be generated for `admin` role and used for API calls.

miltonhultgren replied:
I wonder if we should rename this function to indicate that it returns an admin level client? 
Else I don't see how we'll remember to modify this behaviour when we split the tests apart.

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/agnostic-tests

flash1293 committed:
remove old reference

flash1293 committed:
Merge branch 'flash1293/agnostic-tests' of github.com:flash1293/kibana into flash1293/agnostic-tests

tonyghiani commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/service.ts:
```diff
@@ -52,6 +52,8 @@ export class StreamsService {
 
     const scopedClusterClient = coreStart.elasticsearch.client.asScoped(request);
 
+    const isServerless = coreStart.elasticsearch.getCapabilities().serverless;
```
**doubt**: is this the recommended way to ensure we are in a serverless environment? 
I remember this was also an option and feels more config-based, wdyt?
```ts
const isServerless = pluginContext.env.packageInfo.buildFlavor === 'serverless';
```

flash1293 replied:
By my understanding this is the recommended way when it's about Elasticsearch capabilities (which is the case here - serverless doesn't support failure store yet)

tonyghiani reviewed (approved):
Doubts at [this comment](https://github.com/elastic/kibana/pull/206979/files#r1922130791) solved and the changes LGTM.

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/268324)
* Commit: f06569e983beaf257d190d3ea7527971490e3b9c

### Failed CI Steps
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/268324#019482ea-d197-4de0-a1cf-61689812ca23)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/268324#019482ea-d197-4de0-a1cf-61689812ca23) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/268324/jobs/019482ea-d197-4de0-a1cf-61689812ca23/artifacts/0194831c-0dbd-454c-b420-1274720e6948) Jest Tests #<span></span>17 / GraphInvestigation Component shows error on bad kql syntax

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #268167](https://buildkite.com/elastic/kibana-pull-request/builds/268167) failed dd14269140e717496fa5f0e4da73c5ce3b27b297
* :broken_heart: [Build #268100](https://buildkite.com/elastic/kibana-pull-request/builds/268100) failed a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c
* :broken_heart: [Build #268043](https://buildkite.com/elastic/kibana-pull-request/builds/268043) failed f9de8ed5432889025d6f9ec7af2ebff85ba54aaa
* :broken_heart: [Build #267860](https://buildkite.com/elastic/kibana-pull-request/builds/267860) failed fd658686d2964290ab6c6ca0fb29a4953fa6550c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268324","number":268324,"commit":"f06569e983beaf257d190d3ea7527971490e3b9c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268167","number":268167,"commit":"dd14269140e717496fa5f0e4da73c5ce3b27b297"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268100","number":268100,"commit":"a6bef0f763ba0621b8b2a638f1acc3b4bebdeb5c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268043","number":268043,"commit":"f9de8ed5432889025d6f9ec7af2ebff85ba54aaa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267860","number":267860,"commit":"fd658686d2964290ab6c6ca0fb29a4953fa6550c"}],"number":268324}
buildkite-pr-comment-->

flash1293 committed:
rename

flash1293 commented:
> I wonder if we should rename this function to indicate that it returns an admin level client?
Else I don't see how we'll remember to modify this behaviour when we split the tests apart.

Fair point, I renamed this

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/12869184966
